std/parallelism.d

```diff
@@ -99,86 +99,56 @@ else version(NetBSD)
     version = useSysctlbyname;
 }
 
-
-version(Windows)
-{
-    // BUGS:  Only works on Windows 2000 and above.
-    shared static this()
-    {
-        import core.sys.windows.windows : SYSTEM_INFO, GetSystemInfo;
-        import std.algorithm.comparison : max;
-
-        SYSTEM_INFO si;
-        GetSystemInfo(&si);
-        totalCPUs = max(1, cast(uint) si.dwNumberOfProcessors);
-    }
-
-}
-else version(linux)
-{
-    shared static this()
-    {
-        import core.sys.posix.unistd : _SC_NPROCESSORS_ONLN, sysconf;
-        totalCPUs = cast(uint) sysconf(_SC_NPROCESSORS_ONLN);
-    }
-}
-else version(Solaris)
+/*
+A lazily initialized global constant. The underlying value is a shared global
+statically initialized to `outOfBandValue` which must not be a legit value of
+the constant. Upon the first call the situation is detected and the global is
+initialized by calling `initializer`. The initializer is assumed to be pure
+(even if not marked as such), i.e. return the same value upon repeated calls.
+For that reason, no special precautions are taken so `initializer` may be called
+more than one time leading to benign races on the cached value.
+
+In the quiescent state the cost of the function is an atomic load from a global.
+*/
+package @property @nogc nothrow pure @trusted
+T lazilyInitializedConstant(T, T outOfBandValue, alias initializer)()
 {
-    shared static this()
-    {
-        import core.sys.posix.unistd : _SC_NPROCESSORS_ONLN, sysconf;
-        totalCPUs = cast(uint) sysconf(_SC_NPROCESSORS_ONLN);
-    }
+    static nothrow T impl()
+    {
+        static shared T result = outOfBandValue;
+        // Short path
+        auto local = atomicLoad(result);
+        if (local != outOfBandValue) return local;
+        // Long path
+        local = initializer();
```

Comment:
Use different variable name here, as this use of `local` is completely independent of the previous one.

Reply:
not getting it... it's the same var

Reply:
Consider the trivial case:
```
int i = 4;
foo(i);
i = 5;  // really the start of a different variable
foo(i);
```
By the idea that each variable should have the tightest scope it can, when a variable has multiple disjoint lifetimes it should be split into multiple independent variables. It also often enables a variable to become `const`.

Reply:
@WalterBright I understood the larger point you made with your first comment, but here `local` is and stays the local representation of the data meant to be cached throughout. We really are in good shape here.